cgsmiles/resolve.py

```diff
@@ -69,62 +68,80 @@ def match_bonding_descriptors(source, target, bond_attribute="bonding"):
+    in the CGSmiles string.
     """
-
     def __init__(self,
                  pattern,
                  meta_graph=None,
                  fragment_dict={},
-                 all_atom=True):
+                 last_all_atom=True):
 
-        # let's start by setting some attributes
-        # this is the fragment string
-        self.fragment_str = None
-        # this is the cgsmiles string
-        self.cgsmiles_str = None
+        """
+        Parameters
+        ----------
+        pattern: str
+            the cgsmiles string to resolve
+        meta_graph: `:class:nx.Graph`
+            a potential higher resolution graph
+        fragment_dict: dict[`:class:nx.Garph`]
+            a dict of fragment graphs
+        last_all_atom: bool
+            if the last resolution is at the all
+            atom level. Default: True
+        """
         # this is the dict with the fragments
         # either provided and/or extracted
         # from the fragment string
         self.fragment_dict = fragment_dict
-        # this is the lower resolution graph
-        self.meta_graph = meta_graph
-        # this is the higher resolution graph
+        # this is the current resolution graph
         self.molecule = nx.Graph()
-        # this attribute stores if the fragments
-        # follow open_smiles syntax or cg_smiles
-        # syntax
-        self.all_atom = all_atom
+        # this is the current meta_graph
+        self.meta_graph = nx.Graph()
+        # these are possible fragment strings
+        self.fragment_strs = []
 
-        # here we figure out what we are dealing with
+        # here we figure out how many resolutions
+        # we are dealing with
         elements = re.findall(r"\{[^\}]+\}", pattern)
+
         # case 1)
         # a meta_graph is provided which means we only
         # have fragments to deal with
         if meta_graph:
-            if len(elements) > 1:
-                msg = ("When providing a meta_graph, the pattern can only"
-                       "contain fragment.")
-                raise IOError(msg)
-            else:
-                self.fragment_string = elements[0]
-
+            self.fragment_strs = elements
         # case 2) we have a meta graph only described
         # and the fragment come from elsewhere
         elif len(elements) == 1 and self.fragment_dict:
-            self.cgsmiles_string = elements[0]
-
+            self.molecule = read_cgsmiles(elements[0])
         # case 3) a string containing both fragments and
         # the meta sequence is provided
         else:
-            if len(elements) < 2:
-                msg = ("When providing a meta_graph, the pattern can only"
-                       "contain fragment.")
-                raise IOError(msg)
-            else:
-                self.cgsmiles_string = elements[0]
-                self.fragment_string = elements[1]
+            self.molecule = read_cgsmiles(elements[0])
+            self.fragment_strs = elements[1:]
```

Comment:
Please move the initialization of `self.molecule` and `self.fragment_strs` (to empty defaults) to the relevant branches of this if-else. It'll make understanding the code easier.

Reply:
done, in fact there was actually a logic mistake.